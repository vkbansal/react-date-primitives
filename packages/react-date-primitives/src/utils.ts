export const DayOfMonthSymbol = Symbol('DayOfMonth');
export const DayOfRangeMonthSymbol = Symbol('DayOfRangeMonth');

export enum DayName {
    MONDAY = 'MONDAY',
    TUESDAY = 'TUESDAY',
    WEDNESDAY = 'WEDNESDAY',
    THURSDAY = 'THURSDAY',
    FRIDAY = 'FRIDAY',
    SATURDAY = 'SATURDAY',
    SUNDAY = 'SUNDAY'
}

export const DaysOfTheWeek: DayName[] = [
    DayName.SUNDAY,
    DayName.MONDAY,
    DayName.TUESDAY,
    DayName.WEDNESDAY,
    DayName.THURSDAY,
    DayName.FRIDAY,
    DayName.SATURDAY
];

type Mutable<T> = {
    -readonly [P in keyof T]: T[P];
};

export interface DayOfMonth {
    readonly __type: typeof DayOfMonthSymbol;
    readonly dateObj: Date;
    readonly dayName: DayName;
    readonly inCurrentMonth: boolean;
    readonly ISODateString: string;
}

export interface DayOfRangeMonth extends Omit<DayOfMonth, '__type'> {
    readonly __type: typeof DayOfRangeMonthSymbol;
    readonly inRange: boolean;
    readonly isStart: boolean;
    readonly isEnd: boolean;
}

export interface Month {
    readonly days: DayOfMonth[];
    readonly daysOfWeek: DayName[];
    readonly month: Date;
}

export interface RangeMonth {
    readonly days: DayOfRangeMonth[];
    readonly month: Date;
}

export interface RangeMonths {
    readonly months: RangeMonth[];
    readonly daysOfWeek: DayName[];
}

const monthsCache = new Map<string, Month>();
const rangeMonthsCache = new Map<string, RangeMonths>();

export function addDays(date: Date, days: number): Date {
    return setDay(date, date.getDate() + days);
}

export function addMonths(date: Date, months: number): Date {
    return setMonth(date, date.getMonth() + months);
}

export function setDay(date: Date, day: number): Date {
    const newDate = new Date(date.getTime());

    newDate.setDate(day);

    return newDate;
}

export function setMonth(date: Date, month: number): Date {
    const newDate = new Date(date.getTime());

    newDate.setMonth(month);

    return newDate;
}

export function setYear(date: Date, year: number): Date {
    const newDate = new Date(date.getTime());

    newDate.setFullYear(year);

    return newDate;
}

export function startOfMonth(date: Date): Date {
    const newDate = new Date(date.getTime());

    newDate.setDate(1);
    newDate.setHours(0, 0, 0, 0);

    return newDate;
}

export function endOfMonth(date: Date): Date {
    const newDate = new Date(date.getTime());

    newDate.setFullYear(date.getFullYear(), date.getMonth() + 1, 0);
    newDate.setHours(23, 59, 59, 999);

    return newDate;
}

export function isSameMonth(dateLeft: Date, dateRight: Date): boolean {
    return (
        dateLeft.getFullYear() === dateRight.getFullYear() &&
        dateLeft.getMonth() === dateRight.getMonth()
    );
}

export function isSameDay(dateLeft: Date, dateRight: Date): boolean {
    return (
        dateLeft.getFullYear() === dateRight.getFullYear() &&
        dateLeft.getMonth() === dateRight.getMonth() &&
        dateLeft.getDate() === dateRight.getDate()
    );
}

export function isDayAfter(date: Date, dateToCompare: Date): boolean {
    const newDate = new Date(date.getTime());
    const newDateToCompare = new Date(dateToCompare.getTime());

    newDate.setHours(0, 0, 0, 0);
    newDateToCompare.setHours(0, 0, 0, 0);

    return newDate.getTime() > newDateToCompare.getTime();
}

export function isDayBefore(date: Date, dateToCompare: Date): boolean {
    const newDate = new Date(date.getTime());
    const newDateToCompare = new Date(dateToCompare.getTime());

    newDate.setHours(0, 0, 0, 0);
    newDateToCompare.setHours(0, 0, 0, 0);

    return newDate.getTime() < newDateToCompare.getTime();
}

export function toISODateString(date: Date): string {
    const year = date.getFullYear();
    const month = (date.getMonth() + 1).toString().padStart(2, '0');
    const day = (date.getDate() + 0).toString().padStart(2, '0');

    return `${year}-${month}-${day}`;
}

export function callIfExists<T extends unknown[], U = unknown>(
    func: ((...args: T) => U) | undefined | null,
    ...args: T
): U | undefined {
    if (typeof func === 'function') {
        return func(...args);
    }
}

export function getDaysOfMonth(month: Date, weekStartsOn = DayName.SUNDAY): Month {
    const firstDate = startOfMonth(month);
    const dayOffset = DaysOfTheWeek.findIndex((d) => d === weekStartsOn);
    const key = `${toISODateString(firstDate)}-${weekStartsOn}`;

    if (dayOffset < 0) {
        throw new RangeError(`Invalid value for day provided: ${weekStartsOn}`);
    }

    const data = monthsCache.get(key);

    if (data) {
        return data;
    }

    const firstDayInCurrentMonth = firstDate.getDay();

    const days: Month['days'] = Array.from(
        { length: 6 * 7 },
        (_1, i): DayOfMonth => {
            const day = addDays(
                firstDate,
                i -
                    firstDayInCurrentMonth -
                    (dayOffset > firstDayInCurrentMonth ? 7 : 0) +
                    dayOffset
            );

            return {
                __type: DayOfMonthSymbol,
                dateObj: day,
                dayName: DaysOfTheWeek[(i + dayOffset) % 7],
                inCurrentMonth: isSameMonth(firstDate, day),
                ISODateString: toISODateString(day)
            };
        }
    );

    const daysOfWeek = DaysOfTheWeek.slice(dayOffset).concat(DaysOfTheWeek.slice(0, dayOffset));

    monthsCache.set(key, { days, daysOfWeek, month: firstDate });

    return { days, daysOfWeek, month: firstDate };
}

export function getDaysOfRangeMonth(
    rangeMonths: Date[],
    startDate?: Date | null,
    endDate?: Date | null,
    weekStartsOn = DayName.SUNDAY
): RangeMonths {
    let key = rangeMonths
        .map((month): string => {
            if (!(month instanceof Date)) {
                throw new TypeError('Given month is not an instance of Date');
            }
            return `${month.getFullYear()}-${month.getMonth()}`;
        })
        .join('|');

    key += `|${startDate ? toISODateString(startDate) : ''}`;
    key += `|${endDate ? toISODateString(endDate) : ''}`;
    key += `|${weekStartsOn}`;

    if (rangeMonthsCache.has(key)) {
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        return rangeMonthsCache.get(key)!;
    }

    const months = rangeMonths.reduce<Mutable<RangeMonths>>(
        (accumulator, currentMonth, i) => {
            const { days, daysOfWeek } = getDaysOfMonth(currentMonth, weekStartsOn);

            if (i === 0) {
                accumulator.daysOfWeek = daysOfWeek.slice(0);
            }

            const daysOfCurrentMonth = days.map(
                (day): DayOfRangeMonth => {
                    const isStart = Boolean(startDate && isSameDay(day.dateObj, startDate));
                    const isEnd = Boolean(endDate && isSameDay(day.dateObj, endDate));

                    return {
                        ...day,
                        __type: DayOfRangeMonthSymbol,
                        inRange:
                            isStart ||
                            isEnd ||
                            Boolean(
                                startDate &&
                                    isDayAfter(day.dateObj, startDate) &&
                                    endDate &&
                                    isDayBefore(day.dateObj, endDate)
                            ),
                        isStart,
                        isEnd
                    };
                }
            );

            accumulator.months.push({
                days: daysOfCurrentMonth,
                month: startOfMonth(currentMonth)
            });

            return accumulator;
        },
        { months: [], daysOfWeek: [] }
    );

    rangeMonthsCache.set(key, months);

    return months as RangeMonths;
}
