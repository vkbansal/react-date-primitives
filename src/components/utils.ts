import { DayOfMonthSubset } from './types';

const monthsCache: Map<string, DayOfMonthSubset[][]> = new Map();

export function setDay(date: Date, day: number): Date {
    const newDate = new Date(date.getTime());

    newDate.setDate(day);

    return newDate;
}

export function addDays(date: Date, days: number): Date {
    return setDay(date, date.getDate() + days);
}

export function setMonth(date: Date, month: number): Date {
    const newDate = new Date(date.getTime());

    newDate.setMonth(month);

    return newDate;
}

export function addMonths(date: Date, months: number): Date {
    return setMonth(date, date.getMonth() + months);
}

export function setYear(date: Date, year: number): Date {
    const newDate = new Date(date.getTime());

    newDate.setFullYear(year);

    return newDate;
}

export function startOfMonth(date: Date): Date {
    const newDate = new Date(date.getTime());

    newDate.setDate(1);
    newDate.setHours(0, 0, 0, 0);

    return newDate;
}

export function endOfMonth(date: Date): Date {
    const newDate = new Date(date.getTime());

    newDate.setFullYear(date.getFullYear(), date.getMonth() + 1, 0);
    newDate.setHours(23, 59, 59, 999);

    return newDate;
}

export function isSameMonth(dateLeft: Date, dateRight: Date): boolean {
    return (
        dateLeft.getFullYear() === dateRight.getFullYear() &&
        dateLeft.getMonth() === dateRight.getMonth()
    );
}

export function isSameDay(dateLeft: Date, dateRight: Date): boolean {
    return (
        dateLeft.getFullYear() === dateRight.getFullYear() &&
        dateLeft.getMonth() === dateRight.getMonth() &&
        dateLeft.getDate() === dateRight.getDate()
    );
}

export function isDayAfter(date: Date, dateToCompare: Date): boolean {
    const newDate = new Date(date.getTime());
    const newDateToCompare = new Date(dateToCompare.getTime());

    newDate.setHours(0, 0, 0, 0);
    newDateToCompare.setHours(0, 0, 0, 0);

    return newDate.getTime() > newDateToCompare.getTime();
}

export function isDayBefore(date: Date, dateToCompare: Date): boolean {
    const newDate = new Date(date.getTime());
    const newDateToCompare = new Date(dateToCompare.getTime());

    newDate.setHours(0, 0, 0, 0);
    newDateToCompare.setHours(0, 0, 0, 0);

    return newDate.getTime() < newDateToCompare.getTime();
}

export function callIfExists(callback?: any, ...args: any[]) {
    if (typeof callback === 'function') {
        callback(...args);
    }
}

export function getDatesofMonth(month: Date): DayOfMonthSubset[][] {
    const firstDate = startOfMonth(month);
    const key = `${firstDate.getFullYear()}-${firstDate.getMonth()}`;

    const data = monthsCache.get(key);

    if (data) {
        return data;
    }

    const firstDayInCurrentMonth = firstDate.getDay();

    const newdata: DayOfMonthSubset[][] = Array.from(
        { length: 6 },
        (_1, i): DayOfMonthSubset[] => {
            const week = Array.from(
                { length: 7 },
                (_2, j): DayOfMonthSubset => {
                    const day = addDays(firstDate, i * 7 + j - firstDayInCurrentMonth);

                    return { date: day, inCurrentMonth: isSameMonth(firstDate, day) };
                }
            );

            return week;
        }
    );

    monthsCache.set(key, newdata);

    return newdata;
}
