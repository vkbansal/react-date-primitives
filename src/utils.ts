export interface DayOfMonth {
    date: Date;
    inCurrentMonth: boolean;
}

export type DaysOfMonth = DayOfMonth[][];

const monthsCache: Map<string, DaysOfMonth> = new Map();

export function setDay(date: Date, day: number): Date {
    const newDate = new Date(date.getTime());

    newDate.setDate(day);

    return newDate;
}

export function addDays(date: Date, days: number): Date {
    return setDay(date, date.getDate() + days);
}

export function setMonth(date: Date, month: number): Date {
    const newDate = new Date(date.getTime());

    newDate.setMonth(month);

    return newDate;
}

export function addMonths(date: Date, months: number): Date {
    return setMonth(date, date.getMonth() + months);
}

export function setYear(date: Date, year: number): Date {
    const newDate = new Date(date.getTime());

    newDate.setFullYear(year);

    return newDate;
}

export function startOfMonth(date: Date): Date {
    const newDate = new Date(date.getTime());

    newDate.setDate(1);
    newDate.setHours(0, 0, 0, 0);

    return newDate;
}

export function endOfMonth(date: Date): Date {
    const newDate = new Date(date.getTime());

    newDate.setFullYear(date.getFullYear(), date.getMonth() + 1, 0);
    newDate.setHours(23, 59, 59, 999);

    return newDate;
}

export function isSameMonth(dateLeft: Date, dateRight: Date): boolean {
    return (
        dateLeft.getFullYear() === dateRight.getFullYear() &&
        dateLeft.getMonth() === dateRight.getMonth()
    );
}

export function isSameDay(dateLeft: Date, dateRight: Date): boolean {
    return (
        dateLeft.getFullYear() === dateRight.getFullYear() &&
        dateLeft.getMonth() === dateRight.getMonth() &&
        dateLeft.getDate() === dateRight.getDate()
    );
}

export function isDayAfter(date: Date, dateToCompare: Date): boolean {
    const newDate = new Date(date.getTime());
    const newDateToCompare = new Date(dateToCompare.getTime());

    newDate.setHours(0, 0, 0, 0);
    newDateToCompare.setHours(0, 0, 0, 0);

    return newDate.getTime() > newDateToCompare.getTime();
}

export function isDayBefore(date: Date, dateToCompare: Date): boolean {
    const newDate = new Date(date.getTime());
    const newDateToCompare = new Date(dateToCompare.getTime());

    newDate.setHours(0, 0, 0, 0);
    newDateToCompare.setHours(0, 0, 0, 0);

    return newDate.getTime() < newDateToCompare.getTime();
}

export function callIfExists(callback?: any, ...args: any[]) {
    if (typeof callback === 'function') {
        callback(...args);
    }
}

export function getDaysOfMonth(month: Date): DayOfMonth[][] {
    const firstDate = startOfMonth(month);
    const key = `${firstDate.getFullYear()}-${firstDate.getMonth()}`;

    const data = monthsCache.get(key);

    if (data) {
        return data;
    }

    const firstDayInCurrentMonth = firstDate.getDay();

    const newdata: DaysOfMonth = Array.from(
        { length: 6 },
        (_1, i): DayOfMonth[] => {
            const week = Array.from(
                { length: 7 },
                (_2, j): DayOfMonth => {
                    const day = addDays(firstDate, i * 7 + j - firstDayInCurrentMonth);

                    return { date: day, inCurrentMonth: isSameMonth(firstDate, day) };
                }
            );

            return week;
        }
    );

    monthsCache.set(key, newdata);

    return newdata;
}
