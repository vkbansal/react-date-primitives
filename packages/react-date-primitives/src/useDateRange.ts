import { useState, useCallback } from 'react';
import type { Dispatch, SetStateAction } from 'react';

import { getDaysOfRangeMonth, isDayBefore, DayName, RangeMonths } from './utils';

export type DateRange = [Date | null, Date | null];

export interface UseDateRangeOptions {
  range?: [Date | null, Date | null];
  rangeEndDate?: Date;
  weekStartsOn?: DayName;
}

export interface UseDateRangeReturn extends RangeMonths {
  readonly range: [Date | null, Date | null];
  setMonths: Dispatch<SetStateAction<Date[]>>;
  setRange: Dispatch<SetStateAction<[Date | null, Date | null]>>;
  setStartOfWeek: Dispatch<SetStateAction<DayName>>;
}

export function useDateRange(
  rangeMonths: Date[],
  options: UseDateRangeOptions = {}
): UseDateRangeReturn {
  const [currentMonths, setCurrentMonths] = useState<Date[]>(rangeMonths);
  const [startOfWeek, setStartOfWeek] = useState(options?.weekStartsOn || DayName.SUNDAY);
  const [range, setRange] = useState<DateRange>([
    options?.range?.[0] || null,
    options?.range?.[1] || null
  ]);

  const { months, daysOfWeek } = getDaysOfRangeMonth(
    currentMonths,
    range[0],
    range[1],
    startOfWeek
  );

  const setRangeCallBack = useCallback(
    (newRange: SetStateAction<DateRange>) => {
      setRange((oldRange) => {
        const [start, end] = typeof newRange === 'function' ? newRange(oldRange) : newRange;
        if (start && end && isDayBefore(end, start)) {
          throw new RangeError(`Range end cannot be before the range start`);
        }

        return [start, end];
      });
    },
    [setRange]
  );

  return {
    months,
    daysOfWeek,
    range,
    setMonths: setCurrentMonths,
    setRange: setRangeCallBack,
    setStartOfWeek
  };
}
