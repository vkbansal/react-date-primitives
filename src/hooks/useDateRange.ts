import { useState } from 'react';

import { DayOfMonth, isDayBefore, isDayAfter, isSameDay } from '../utils';

export interface DayOfRangeMonth extends DayOfMonth {
    inRange: boolean;
    selected: boolean;
    isStart: boolean;
    isEnd: boolean;
}

export interface UseDateRange {
    handleDayClick(day: Date): void;
    handleDayHover(day: Date): void;
    startDate: Date;
    endDate?: Date;
    processMonth(month: DayOfMonth[][]): DayOfRangeMonth[][];
}

export function useDateRange(
    initialStartDate: Date = new Date(),
    initialEndDate?: Date
): UseDateRange {
    const [startDate, setStartDate] = useState(initialStartDate);
    const [endDate, setEndDate] = useState(initialEndDate);
    const [selectionActive, setSelectionActive] = useState(false);

    function handleDayClick(day: Date) {
        if (startDate && selectionActive && !isDayBefore(day, startDate)) {
            setStartDate(startDate);
            setEndDate(day);
            setSelectionActive(false);
        } else {
            setStartDate(day);
            setEndDate(undefined);
            setSelectionActive(true);
        }
    }

    function handleDayHover(day: Date) {
        if (selectionActive && startDate && !isDayBefore(day, startDate)) {
            setEndDate(day);
        }
    }

    function processMonth(month: DayOfMonth[][]): DayOfRangeMonth[][] {
        return month.map<DayOfRangeMonth[]>((week) =>
            week.map<DayOfRangeMonth>((day) => {
                const isStart = startDate && isSameDay(day.date, startDate);
                const isEnd = Boolean(endDate && isSameDay(day.date, endDate));

                return {
                    ...day,
                    selected: isStart || isEnd,
                    inRange: Boolean(
                        startDate &&
                            isDayAfter(day.date, startDate) &&
                            endDate &&
                            isDayBefore(day.date, endDate)
                    ),
                    isStart,
                    isEnd
                };
            })
        );
    }

    return { handleDayClick, handleDayHover, startDate, endDate, processMonth };
}
