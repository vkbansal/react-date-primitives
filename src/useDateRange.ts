import { useState, SetStateAction, Dispatch } from 'react';

import { getDaysOfRangeMonth, isSameDay, isDayAfter, DayOfWeek, RangeMonths } from './utils';

export interface UseDateRangeOptions {
    rangeStartDate?: Date;
    rangeEndDate?: Date;
    weekStartsOn?: DayOfWeek;
}

export interface DateRange extends RangeMonths {
    readonly startDate: Date | null;
    readonly endDate: Date | null;
    setMonths: Dispatch<SetStateAction<Date[]>>;
    setStartDate: Dispatch<SetStateAction<Date | null>>;
    setEndDate: Dispatch<SetStateAction<Date | null>>;
}

export function useDateRange(rangeMonths: Date[], options: UseDateRangeOptions = {}): DateRange {
    const [currentMonths, setCurrentMonths] = useState<Date[]>(rangeMonths);
    const [startDate, setStartDate] = useState<Date | null>(options.rangeEndDate || null);
    const [endDate, setEndDate] = useState<Date | null>(options.rangeEndDate || null);

    const { months, daysOfWeek } = getDaysOfRangeMonth(currentMonths, startDate, endDate);

    return {
        months,
        daysOfWeek,
        startDate,
        endDate,
        setMonths: setCurrentMonths,
        setStartDate(date: SetStateAction<Date | null>): void {
            setStartDate(date);
            setEndDate(null);
        },
        setEndDate(date: SetStateAction<Date | null>): void {
            if (
                startDate &&
                date instanceof Date &&
                (isSameDay(date, startDate) || isDayAfter(date, startDate))
            ) {
                setEndDate(date);
            } else if (typeof date === 'function') {
                setEndDate(date);
            }
        }
    };
}
