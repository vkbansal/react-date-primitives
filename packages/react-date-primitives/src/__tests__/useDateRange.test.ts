import { renderHook, act } from '@testing-library/react-hooks';

import { useDateRange } from '../useDateRange';
import { isSameDay, addMonths } from '../utils';

import DateSerializer from './DateSerializer';
import MonthSerializer from './MonthSerializer';

expect.addSnapshotSerializer(DateSerializer);
expect.addSnapshotSerializer(MonthSerializer);

const YEAR = 2021;

describe('useDateRange hook tests', () => {
  const month = new Date(YEAR, 0 /* Jan */, 1, 0, 0, 0, 0);
  const months = Array.from({ length: 12 }, (_, i) => addMonths(month, i));

  test('returns basic properties', () => {
    const { result } = renderHook(() => useDateRange(months.slice(0, 3)));

    expect(result.current.range).toEqual([null, null]);
    expect(Array.isArray(result.current.months)).toBe(true);
    expect(result.current.months.every((month) => Array.isArray(month.days))).toBe(true);
    expect(typeof result.current.setRange).toBe('function');
    expect(typeof result.current.setMonths).toBe('function');
    expect(result.current.months[0].days[0]).toMatchInlineSnapshot(`
            Object {
              "ISODateString": "2020-12-27",
              "__type": Symbol(DayOfRangeMonth),
              "dateObj": 2020-12-27,
              "dayName": "SUNDAY",
              "inCurrentMonth": false,
              "inRange": false,
              "isEnd": false,
              "isStart": false,
            }
        `);
  });

  test('setRange works when endDate > startDate', () => {
    const date1 = new Date(YEAR, 0 /* Jan */, 1, 0, 0, 0, 0);
    const date2 = new Date(YEAR, 0 /* Jan */, 9, 0, 0, 0, 0);
    const { result } = renderHook(() => useDateRange(months.slice(0, 3)));

    expect(result.current.range).toEqual([null, null]);
    expect(result.current.months[0]).toMatchSnapshot('1. init');

    // set a startDate
    act(() => {
      result.current.setRange([date1, null]);
    });
    expect(isSameDay(result.current.range[0]!, date1)).toBe(true);
    expect(result.current.range[1]).toBeNull();

    expect(result.current.months[0]).toMatchSnapshot('2. starDate is set');

    // set an endDate > startDate
    act(() => {
      result.current.setRange(() => [date1, date2]);
    });
    expect(isSameDay(result.current.range[0]!, date1)).toBe(true);
    expect(isSameDay(result.current.range[1]!, date2)).toBe(true);

    expect(result.current.months[0]).toMatchSnapshot('3. endDate is set');
  });

  test('setRange throws when endDate < startDate', () => {
    try {
      const date1 = new Date(YEAR, 0 /* Jan */, 1, 0, 0, 0, 0);
      const date2 = new Date(YEAR, 0 /* Jan */, 9, 0, 0, 0, 0);
      const { result } = renderHook(() => useDateRange(months.slice(0, 3)));

      // set a startDate
      act(() => {
        result.current.setRange([date2, null]);
      });
      expect(isSameDay(result.current.range[0]!, date2)).toBe(true);
      expect(result.current.range[1]).toBeNull();

      // set an endDate < startDate
      act(() => {
        result.current.setRange([date2, date1]);
      });

      expect(result.error).toEqual(RangeError('Range end cannot be before the range start'));
    } catch (e) {
      //
    }
  });
});
