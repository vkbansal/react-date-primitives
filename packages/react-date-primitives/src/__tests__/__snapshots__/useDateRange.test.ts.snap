// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useDateRange hook tests setRange works when endDate > startDate: 1. init 1`] = `
|                Jan 2021                 |
|-----------------------------------------|
|
|     |     |     |     |     |  01 |  02 |
|  03 |  04 |  05 |  06 |  07 |  08 |  09 |
|  10 |  11 |  12 |  13 |  14 |  15 |  16 |
|  17 |  18 |  19 |  20 |  21 |  22 |  23 |
|  24 |  25 |  26 |  27 |  28 |  29 |  30 |
|  31 |     |     |     |     |     |     |
`;

exports[`useDateRange hook tests setRange works when endDate > startDate: 2. starDate is set 1`] = `
|                Jan 2021                 |
|-----------------------------------------|
|
|     |     |     |     |     |  ** |  02 |
|  03 |  04 |  05 |  06 |  07 |  08 |  09 |
|  10 |  11 |  12 |  13 |  14 |  15 |  16 |
|  17 |  18 |  19 |  20 |  21 |  22 |  23 |
|  24 |  25 |  26 |  27 |  28 |  29 |  30 |
|  31 |     |     |     |     |     |     |
`;

exports[`useDateRange hook tests setRange works when endDate > startDate: 3. endDate is set 1`] = `
|                Jan 2021                 |
|-----------------------------------------|
|
|     |     |     |     |     |  ** |  ** |
|  ** |  ** |  ** |  ** |  ** |  ** |  ** |
|  10 |  11 |  12 |  13 |  14 |  15 |  16 |
|  17 |  18 |  19 |  20 |  21 |  22 |  23 |
|  24 |  25 |  26 |  27 |  28 |  29 |  30 |
|  31 |     |     |     |     |     |     |
`;
